crates/exex/exex/src/dyexex.rs

```diff
@@ -0,0 +1,159 @@
+//! Type-safe abstractions for Dynamically Loaded ExExes
+
+use std::{
+    env::consts::{DLL_PREFIX, DLL_SUFFIX},
+    fmt::Debug,
+    future::Future,
+    path::Path,
+    pin::Pin,
+    sync::Arc,
+};
+
+use eyre::{OptionExt, Result};
+use libloading::{Library, Symbol};
+
+use crate::ExExContextDyn;
+
+/// ExEx launch function dynamic library symbol name.
+const LAUNCH_EXEX_FN: &[u8] = b"_launch_exex";
+
+/// Dynamically loads an ExEx entrypoint, which accepts a user-defined function representing the
+/// core ExEx logic. The provided function must take an [`ExExContextDyn`](`crate::ExExContextDyn`)
+/// as its argument.
+///
+/// # Returns
+/// A Future that will be polled by the [`ExExManager`](`crate::ExExManager`).
+///
+/// ## Example usage:
+/// ```rust
+/// use reth_exex::{define_exex, ExExContextDyn};
+/// use reth_node_api::FullNodeComponents;
+/// use std::future::Future;
+///
+/// // Create a function to produce ExEx logic
+/// async fn exex(_ctx: ExExContextDyn) -> eyre::Result<impl Future<Output = eyre::Result<()>>> {
+///     let _exex = async move { Ok(()) };
+///     Ok(_exex)
+/// }
+///
+/// // Use the macro to generate the entrypoint function
+/// define_exex!(exex);
+/// ```
+#[macro_export]
+macro_rules! define_exex {
+    ($user_fn:ident) => {
+        #[no_mangle]
+        pub extern "Rust" fn _launch_exex(
+            ctx: $crate::ExExContextDyn,
+        ) -> futures::future::BoxFuture<
+            'static,
+            eyre::Result<impl std::future::Future<Output = eyre::Result<()>> + Send>,
+        > {
+            Box::pin($user_fn(ctx))
+        }
+    };
+}
+
+#[derive(Debug, Default)]
+/// Dynamic ExEx loader
+pub struct DyExExLoader {
+    /// List of loaded Dynamically Loaded ExExes
+    pub loaded: Vec<LoadedExEx>,
+}
+
+impl DyExExLoader {
+    /// Initializes a Dynamic ExEx loader.
+    /// TODO(0xurb) - path to datadir here and loaded list with capacity of dir inner files.
+    pub const fn new() -> Self {
+        Self { loaded: Vec::new() }
+    }
+
+    /// Loads a dynamically loaded ExEx from a given path.
+    ///
+    /// # Safety
+    ///
+    /// The dynamically loaded ExEx library **must** contains a function with
+    /// name `_launch_exex`.
```

Comment:
the safety note should rather be about the `_launch_exex` function signature, not the naming of the function itself https://docs.rs/libloading/latest/libloading/struct.Library.html#safety-1